src/ol/interaction/modifyinteraction.js

```diff
@@ -959,6 +1044,12 @@ ol.interaction.Modify.prototype.insertVertex_ = function(segmentData, vertex) {
       coordinates = geometry.getCoordinates();
       coordinates[depth[0]].splice(index + 1, 0, vertex);
       break;
+    case ol.geom.GeometryType.RECTANGLE:
```

Comment:
It's how it works for circles. You can just calculate the position of the temporary vertex on handling the mouse down event, and during dragging.